src/csharp/Grpc.Core/SerializationContext.cs

```diff
@@ -0,0 +1,34 @@
+#region Copyright notice and license
+
+// Copyright 2018 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+namespace Grpc.Core
+{
+    /// <summary>
+    /// Provides storage for payload when serializing a message.
+    /// </summary>
+    public abstract class SerializationContext
+    {
+        /// <summary>
+        /// Use the byte array as serialized form of current message and mark serialization process as complete.
+        /// Complete() can only be called once. By calling this method the caller gives up the ownership of the
+        /// payload which must not be accessed afterwards.
+        /// </summary>
+        /// <param name="payload">the serialized form of current message</param>
+        public abstract void Complete(byte[] payload);
```

Comment:
Sorry meant to post this question earlier: what are the restrictions on how `Complete` is called? I'd assume that `Complete` <i>must</i> be called within the contextualSerializer invocation - is there a way to enforce this more strictly if that's true? For example, an implementation couldn't call `Complete` asynchronously from a different thread, right?

Reply:
The assumption is that Complete() must be called within before the serializer returns in order for serialization to be successful. Normally this would would be done from the same thread by the serializer delegate. Not sure yet if calling Complete() from a different thread (if properly synchronized by user code) should be illegal. I will clarify the docs once things are clearer - right now I'm mostly focusing on deserialization.